cli/npm/dsl.d.ts

```diff
@@ -0,0 +1,356 @@
+type AliasRule = {type: 'ALIAS'; named: boolean; content: Rule; value: string};
+type BlankRule = {type: 'BLANK'};
+type ChoiceRule = {type: 'CHOICE'; members: Rule[]};
+type FieldRule = {type: 'FIELD'; name: string; content: Rule};
+type ImmediateTokenRule = {type: 'IMMEDIATE_TOKEN'; content: Rule};
+type PatternRule = {type: 'PATTERN'; value: string};
+type PrecDynamicRule = {type: 'PREC_DYNAMIC'; content: Rule; value: number};
+type PrecLeftRule = {type: 'PREC_LEFT'; content: Rule; value: number};
+type PrecRightRule = {type: 'PREC_RIGHT'; content: Rule; value: number};
+type PrecRule = {type: 'PREC'; content: Rule; value: number};
+type Repeat1Rule = {type: 'REPEAT1'; content: Rule};
+type RepeatRule = {type: 'REPEAT'; content: Rule};
+type SeqRule = {type: 'SEQ'; members: Rule[]};
+type StringRule = {type: 'STRING'; value: string};
+type SymbolRule<Name extends string> = {type: 'SYMBOL'; name: Name};
+type TokenRule = {type: 'TOKEN'; content: Rule};
+
+type Rule =
+  | AliasRule
+  | BlankRule
+  | ChoiceRule
+  | FieldRule
+  | ImmediateTokenRule
+  | PatternRule
+  | PrecDynamicRule
+  | PrecLeftRule
+  | PrecRightRule
+  | PrecRule
+  | Repeat1Rule
+  | RepeatRule
+  | SeqRule
+  | StringRule
+  | SymbolRule<string>
+  | TokenRule;
+
+type RuleOrLiteral = Rule | RegExp | string;
+
+type GrammarSymbols<RuleName extends string> = {
+  [name in RuleName]: SymbolRule<name>;
+} &
```

Comment:
Saying any string key of `$` will be a `SymbolRule` guarantees that prior grammars remain correct and is accurate to the how the `RuleBuilder` works.  However, allowing access to any key means typescript won't be able to catch the invalid reference `$.foobar`.  Does type inference work without the `& Record<string, SymbolRule<string>>`?

Reply:
Correct, the `Record<string, SymbolRule<string>>` is modelling the fact that `$` is a Proxy object, and is required specifically for `externals`. Consider this case:

```typescript
grammar({
  externals: ($) => [$._external],
  rules: {
    example: ($) => choice($._external),
  },
})
```

Without the `& Record<...>`, the only key that TypeScript would be able to infer would be `"example"`, and so the `$._external` would be a type error. 

There is unfortunately not a way for TypeScript to infer the usage of `$.name` inside the `externals` builder function, as there isn't currently a way to model an object in TypeScript where its keys return the string literal type of themselves (see https://github.com/microsoft/TypeScript/issues/22509 for specifically that request). 

If that _did_ exist, then we could write the type such that it inferred the names of all rules + all base grammar rules + the externals, but unfortunately it's just not possible. So today, the `& Record<...>` workaround is probably the best we can do.

If the DSL itself was changed in the future, you could do this:

```typescript
grammar({
  externals: () => ['_external'],
})

// or even just
grammar({
  externals: ['_external'],
})
```

And then it would be possible to infer the names exactly. But that's not compatible with current usage.